lib/3scale/backend/metric.rb

```diff
@@ -149,6 +149,25 @@ def children(service_id, id)
           hierarchy(service_id, false)[id.to_s]
         end
 
+        # Returns the "descendants" of a metric, that is, its children,
+        # grandchildren, etc. in the metric hierarchy of the given service.
+        def descendants(service_id, metric_name)
+          metric_hierarchy = hierarchy(service_id)
+
+          res = metric_hierarchy[metric_name].dup || []
+          pending = res.dup
+
+          until pending.empty?
+            metric = pending.shift
+            children = metric_hierarchy[metric] || []
+
+            res += children
+            pending += children
+          end
```

Comment:
I don't agree. I think that my solution clearly separates the result and the queue of pending metrics. It has more lines, yes, but I think it's easier to understand.
In the solutions that you proposed you are iterating over a growing array directly or via pointers. I find that more difficult to reason about.

Reply:
Regarding your first solution. I also thought about a recursive solution. I thought that it should not be a problem given the number of levels that we should expect, but was not too sure. If you think that it's not a problem. I think that something like this is better because it clearly expresses the idea that the descendants of a metrics are its children plus the descendants of each child:

```ruby
        def descendants(service_id, metric_name)
          metrics_hierarchy = hierarchy(service_id)
          children = metrics_hierarchy[metric_name] || []

          children.reduce(children) do |acc, child|
            acc + descendants(service_id, child)
          end
        end
```

Reply:
My concern is not the "clearly expresses" part but the performance - recursion is probably never going to be very deep, but the overhead of that in performance should be controlled (ie. you are adding a big K in time). Plus we should use `Array#push` rather than `+` to avoid duplicating arrays while dropping temporary ones all the time.

Reply:
I think we shouldn't worry _that_ much about performance until we demonstrate that there's a problem using profiling tools. That unless there's an obvious duplication of work, O(n^2) vs O(n), etc, of course. In this case I don't see anything worrying. If I changed the `+` for `push`, would you find that solution acceptable? You proposed a recursive solution, but reading your latest comment is not clear to me whether you'd accept it or not.

Reply:
We can't measure everything, so we need to develop some sense about what is performant and what is not. The sentence "not worrying (much) about performance until we demonstrate an issue" is featured in many horror tales ending with "and that's why we rewrote the whole thing". :P

My sense of performance says this is not `O(kN)` but `O(Kn)` with a big `K`, and the recursive function not being tail-call means an extra penalty (but I think Ruby does not even optimize it if it was).

I'd accept the recursive version with `Array#push` with the caveat that the next guy that will touch this code will not realize that it is recursive when we finally have arbitrary depths, so at least this would warrant a big note in case someone comes back at you saying "this crashed in production and you didn't warn me!".

Reply:
For completeness, here's a quick benchmark:
```
Comparison:
                each:  2163979.4 i/s
                orig:  1428006.8 i/s - 1.52x  slower
                 rec:  1317247.2 i/s - 1.64x  slower

Benchmarks finished
```
The first being the `.each` solution, the second the original PR and the last the fully recursive version (all of them using `Array#push`).

Not _that_ important unless this sits in the hot path of auths.

Reply:
It might be ~1.5x slower but we don't really know if that is a problem.
What if this method represents only ~0.00...% of the total CPU time?

Reply:
What if it does not? What if it does represent a lot in certain cases with lots of metrics? :D
Also, why should one piece of code be slow just because every other place is slow? :/

Death by a thousand cuts, ok, you win now - add it already!

Reply:
That's what I mentioned several comments above that we need profiling to be sure about this kind of things.

Changed.